### PR TITLE
Include analysis in the UI

### LINKED ISSUE #8: Implement summary statistics for provisioning runs
@djmitche commented on [Thu Aug 06 2020](https://github.com/taskcluster/taskcluster/issues/3360)

* Build a framework for calculating summary stats over a DataStore
* Build a reporting framework in the command-line tools (related to taskcluster/taskcluster#3320) that includes summary stats
* Implement some summary stats
  * overprovisioned_workers -- workers that ran 0 jobs
  * overprovisioned_hours -- hours that those workers ran

---

@djmitche commented on [Thu Aug 06 2020](https://github.com/taskcluster/taskcluster/issues/3360#issuecomment-670233446)

I have this half-finished.

### PR DESCRIPTION
**Pull request by djmitche**
This mostly fixes #8.  There's an endless stream of tweaks to make, but let's call this good for the moment.

### CODE REVIEW

**Comment by djmitche** [comment]
By the way, thoughts for the future here:
 * It'd be really useful to have tests for this analysis stuff.  Most of it is pretty simple, but I had to futz with the overprovisioned workers stuff a bit to get it right, and testing-by-running is not very efficient.  Currently the code is JSX and uses react hooks, so it's difficult to test without webpack and react infrastructure set up.
 * I expect this kind of analysis will eventually be slow, and doing it synchronously in the browser will become problematic.  It'd be great if we could build a framework to do the analysis in a worker thread and show a spinner until each part of the analysis is complete.
 * We'll want to be able to send links to people to show them this-or-that analysis.  Some ideas to that end:
   * Deploy the UI to Netlify
   * Take a URL to a datastore as a URL argument (along with the other URL components detailing which view to show)
   * Run simulations in tasks, with a nice local command to set them up, and producing the datastore as a URL and a nice link to the analysis tool containing that URL.

I may be getting carried away :)

File: package.json [hunk 1]
```diff
@@ -16,11 +16,13 @@
     "chalk": "^4.1.0",
     "commander": "^6.0.0",
     "eslint": "^7.6.0",
+    "moment": "^2.27.0",
```
Comment (helfi92) [inline package.json]: `moment` is a heavy package for the UI. Thoughts on using `date-fns` instead? https://date-fns.org/docs/Getting-Started

File: src/ui/App.jsx [hunk 2]
```diff
@@ -1,17 +1,31 @@
 import { hot } from 'react-hot-loader';
+import {
+  HashRouter as Router,
```
Comment (helfi92) [inline src/ui/App.jsx]: What was the idea behind using `HashRouter` for the routing rather than `BrowserRouter`?

File: src/ui/analysis.jsx [hunk 3]
```diff
@@ -0,0 +1,125 @@
+import React from 'react';
+import moment from 'moment';
+import DataStoreContext from './datastore';
+
+/**
+ * A singleton class for calculating analysis of a run.
+ *
+ * This is full of `get` methods, which use React hooks to ensure that long calculations are
+ * re-done only when necessary.  All of these functions should only be called within a
+ * render function.
```
Comment (helfi92) [inline src/ui/analysis.jsx:10]: What is a render function in react hooks lingo?

File: src/ui/analysis.jsx [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    return React.useMemo(
+      () => {
+        let count = 0;
+        let time = 0;
+        
+        // we only need to look at workers that have shut down, as running
+        // workers may still yet take a task; and only look at resolved tasks
+        // as a task cannot be running on a shut-down worker.
+        for (let {startedTime, shutdownTime, resolvedTasks} of finalState.shutdownWorkers.values()) {
+          console.log(resolvedTasks.size);
```
Comment (helfi92) [inline src/ui/analysis.jsx]: `console.log`

File: src/ui/analysis.jsx [hunk 5]
```diff
@@ -0,0 +1,125 @@
+import React from 'react';
+import moment from 'moment';
+import DataStoreContext from './datastore';
+
+/**
+ * A singleton class for calculating analysis of a run.
```
Comment (helfi92) [inline src/ui/analysis.jsx:6]: Importing this class seems to create a new instance every time. Is it really a singleton?

**Review by helfi92 — COMMENTED** [review]
Looks good! Just minor comments.

File: package.json (hunk 1 above)
Comment (helfi92) [reply package.json]: https://bundlephobia.com/result?p=moment@2.27.0
https://bundlephobia.com/result?p=date-fns@2.15.0

**Review by helfi92 — COMMENTED** [review]
Running the simple simulation throws for me:

```
➜  taskcluster-provisioning-simulator git:(issue8) ✗ yarn sim simple
yarn run v1.21.1
warning ../package.json: No license field
$ node main.js sim simple
➤ Running Simulation
1970-01-01T00:00:00.000Z - ramping up
1970-01-01T00:00:09.000Z - TickTockLoadGenerator - creating task Q0wANbm8RQ6e4WJBW8TPMA
1970-01-01T00:00:09.000Z - w-10000 - requested
1970-01-01T00:00:10.000Z - running simulation
1970-01-01T00:00:18.000Z - TickTockLoadGenerator - creating task FSXyjfshR8yz2OTJmPlyyA
1970-01-01T00:00:18.000Z - w-10001 - requested
1970-01-01T00:00:20.000Z - ramping down
1970-01-01T00:00:29.000Z - w-10000 - started
1970-01-01T00:00:29.000Z - w-10002 - requested
1970-01-01T00:00:29.000Z - w-10000 - claimed Q0wANbm8RQ6e4WJBW8TPMA
1970-01-01T00:00:34.000Z - w-10000 - finished Q0wANbm8RQ6e4WJBW8TPMA
1970-01-01T00:00:34.000Z - w-10000 - claimed FSXyjfshR8yz2OTJmPlyyA
1970-01-01T00:00:38.000Z - w-10001 - started
1970-01-01T00:00:39.000Z - w-10000 - finished FSXyjfshR8yz2OTJmPlyyA
assert.js:100
  throw new AssertionError(obj);
  ^

AssertionError [ERR_ASSERTION]: Expected values to be loosely deep-equal:

[
  'w-10000',
  'w-10001',
  'w-10002'
]

should loosely deep-equal

[]
    at SimpleEstimateProvisioner.stop (/Users/hassanali/Documents/Mozilla/projects/taskcluster-provisioning-simulator/src/provisioner.js:37:12)
    at SimpleSimulator.run (/Users/hassanali/Documents/Mozilla/projects/taskcluster-provisioning-simulator/src/simulator.js:38:22)
    at Command.<anonymous> (/Users/hassanali/Documents/Mozilla/projects/taskcluster-provisioning-simulator/main.js:28:17)
    at Command.listener [as _actionHandler] (/Users/hassanali/Documents/Mozilla/projects/taskcluster-provisioning-simulator/node_modules/commander/index.js:422:31)
    at Command._parseCommand (/Users/hassanali/Documents/Mozilla/projects/taskcluster-provisioning-simulator/node_modules/commander/index.js:975:14)
    at Command._dispatchSubcommand (/Users/hassanali/Documents/Mozilla/projects/taskcluster-provisioning-simulator/node_modules/commander/index.js:926:18)
    at Command._parseCommand (/Users/hassanali/Documents/Mozilla/projects/taskcluster-provisioning-simulator/node_modules/commander/index.js:943:12)
    at Command._dispatchSubcommand (/Users/hassanali/Documents/Mozilla/projects/taskcluster-provisioning-simulator/node_modules/commander/index.js:926:18)
    at Command._parseCommand (/Users/hassanali/Documents/Mozilla/projects/taskcluster-provisioning-simulator/node_modules/commander/index.js:943:12)
    at Command.parse (/Users/hassanali/Documents/Mozilla/projects/taskcluster-provisioning-simulator/node_modules/commander/index.js:778:10) {
  generatedMessage: true,
  code: 'ERR_ASSERTION',
  actual: [Array],
  expected: [],
  operator: 'deepEqual'
}
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

File: src/ui/App.jsx [hunk 6]
```diff
@@ -1,17 +1,31 @@
...
+  Switch,
+  Route,
+  Link
```
Comment (helfi92) [inline src/ui/App.jsx]: `Link` seems to be unused here.

File: src/ui/App.jsx [hunk 7]
```diff
@@ -1,17 +1,31 @@
...
+} from "react-router-dom";
 import React from 'react';
 import './App.css';
-import datastore from '../../datastore';
+import Home from './views/Home';
+import Summary from './views/Summary';
+import DataStoreContext from './datastore';
```
Comment (helfi92) [inline src/ui/App.jsx]: `DataStoreContext` is unused in this file.

File: src/ui/views/Summary.jsx [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import { Link } from "react-router-dom";
+import { hot } from 'react-hot-loader';
+import analysis from '../analysis';
+
+const statistics = [];
```
Comment (helfi92) [inline src/ui/views/Summary.jsx]: Unused.

File: src/ui/views/Summary.jsx [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import { Link } from "react-router-dom";
```
Comment (helfi92) [inline src/ui/views/Summary.jsx]: Unused.

File: src/ui/App.jsx (hunk 2 above)
Comment (djmitche) [reply src/ui/App.jsx]: I sort of assumed BrowserRouter wouldn't work, but I didn't actually check.. :)

File: src/ui/analysis.jsx (hunk 3 above)
Comment (djmitche) [reply src/ui/analysis.jsx:10]: It's the same as non-hooks lingo -- either `render()` on a component class or, for a pure functional component, the function itself.

```javascript
const MyCompoent = (props) => {
  // call hooks stuff in here
};
```

File: src/ui/analysis.jsx (hunk 5 above)
Comment (djmitche) [reply src/ui/analysis.jsx:6]: I add a `console.log` before the `export default`, and it only gets called once.  That matches my understanding of how importing / requiring works -- a module's definition is only executed once and the result cached by the require machinery.  It's similar to `export default () => {..}` -- every place you `import` or `require` that module will get the identical function object.

**Comment by djmitche** [comment]
The thrown error was because I modified the simple sim in such a way that it didn't shut down all of its workers in the ramp-down phase.  I had been fiddling around trying to get it to generate overprovisioned workers.  I reverted my changes to the simulation.

File: .nvmrc [hunk 10]
```diff
@@ -1 +1 @@
-12
+8
```
Comment (djmitche) [inline .nvmrc]: I think I know why this keeps sneaking in.. I'll fix.

File: src/ui/analysis.jsx (hunk 5 above)
Comment (helfi92) [reply src/ui/analysis.jsx:6]: You are right here. My mistake.

File: src/ui/analysis.jsx (hunk 3 above)
Comment (helfi92) [reply src/ui/analysis.jsx:10]: From https://reactjs.org/docs/hooks-faq.html#should-i-use-hooks-classes-or-a-mix-of-both, it looks like class components are not encouraged so render() methods won't be required anymore.

**Review by helfi92 — APPROVED** [review]
Looks great!

File: src/ui/analysis.jsx (hunk 3 above)
Comment (djmitche) [reply src/ui/analysis.jsx:10]: That sounds good to me!  Hooks are actually quite nice once you get used to them.  Is "render function" the right terminology, then?

Comment (helfi92) [reply src/ui/analysis.jsx:10]: It's basically just the return value of the function component. I'm not sure if I would call it render function but I guess that's better.

**Pull request merged**